pkg/finality-grandpa/voting_round.go

```diff
@@ -0,0 +1,842 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"time"
+
+	"golang.org/x/exp/constraints"
+)
+
+type stateStart[T any] [2]T
+
+type stateProposed[T any] [2]T
+
+type statePrevoting[T, W any] struct {
+	T T
+	W W
+}
+
+type statePrevoted[T any] [1]T
+
+type statePrecommitted struct{}
+
+type states[T, W any] interface {
+	stateStart[T] | stateProposed[T] | statePrevoting[T, W] | statePrevoted[T] | statePrecommitted
+}
+
+// The state of a voting round.
+type state any
+
+func setState[T, W any, V states[T, W]](s *state, val V) {
+	*s = val
+}
+
+func newState[T, W any, V states[T, W]](val V) state {
+	var s state
+	setState[T, W](&s, val)
+	return s
+}
+
+type hashBestChain[Hash comparable, Number constraints.Unsigned] struct {
+	Hash      Hash
+	BestChain BestChain[Hash, Number]
+}
+
+// Whether we should vote in the current round (i.e. push votes to the sink.)
+type voting uint
+
+const (
+	// Voting is disabled for the current round.
+	votingNo voting = iota
+	// Voting is enabled for the current round (prevotes and precommits.)
+	votingYes
+	// Voting is enabled for the current round and we are the primary proposer
+	// (we can also push primary propose messages).
+	votingPrimary
+)
```

Comment:
How about just No, Yes and Primary?

Reply:
I feel like private types of `no` and `yes` would be confusing.  I'd rather be more explicit and these are private constants as well.